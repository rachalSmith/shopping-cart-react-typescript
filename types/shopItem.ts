export type itemCategory = "men's clothing" | "women's clothing";

interface IShopItemRating {
  rating: number;
  count: number;
}

export interface IShopItem {
  category: itemCategory;
  description: string;
  id: number;
  image: string;
  price: number;
  rating: IShopItemRating;
  title: string;
}
