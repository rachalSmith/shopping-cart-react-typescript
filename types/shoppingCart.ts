export interface ICartItemRef {
  quantity: number;
  id: number;
}
