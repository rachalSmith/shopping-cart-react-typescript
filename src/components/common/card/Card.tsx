import { ReactNode } from "react";

import { styled } from "@mui/material";
import MuiCard from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import CardMedia from "@mui/material/CardMedia";

import { IShopItem } from "../../../../types/shopItem";

interface ICardProps {
  item: IShopItem;
  children: ReactNode;
  orientation: "column" | "row";
  quickAddButton?: ReactNode;
}

const StyledCardContent = styled(CardContent)({
  paddingTop: "16px",
  height: "80px",
  display: "flex",
  flexDirection: "column",
  justifyContent: "space-between",
  alignItems: "baseline",
  ".MuiCardActionArea-focusHighlight": {
    background: "transparent",
  },
});

const StyledCardMedia = styled(CardMedia)({
  display: "flex",
  justifyContent: "center",
  alignItems: "flex-end",
  objectFit: "scale-down",
  aspectRatio: "2/2.7",
  backgroundPosition: "50% 50%",
  backgroundRepeat: "no-repeat",
  backgroundSize: "cover",
  border: "yellow solid",
});

const Card = ({ item, quickAddButton, children, orientation }: ICardProps) => {
  return (
    <>
      <MuiCard
        variant='outlined'
        sx={{
          border: "red solid",
          display: "flex",
          flexDirection: orientation,
        }}
      >
        <StyledCardMedia image={item.image} title={item.title}>
          {quickAddButton && quickAddButton}
        </StyledCardMedia>

        <StyledCardContent>{children}</StyledCardContent>
      </MuiCard>
    </>
  );
};

export default Card;
