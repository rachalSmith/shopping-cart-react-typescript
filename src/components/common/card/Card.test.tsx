import renderer from "react-test-renderer";

import Card from "./Card";
import { IShopItem } from "../../../../types/shopItem";

const item: IShopItem = {
  category: "men's clothing",
  description: "description",
  id: 1,
  image: "test.png",
  price: 10,
  rating: { rating: 5, count: 10 },
  title: "title",
};

const children = <div>child</div>;
const button = <button>button</button>;

describe("Card component", () => {
  it("should render the card image and children in a row when orientation is 'row'", () => {
    const tree = renderer
      .create(
        <Card item={item} orientation='row' elevation={2}>
          {children}
        </Card>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should render the card image and content in a column when orientation is 'column'", () => {
    const tree = renderer
      .create(
        <Card item={item} orientation='column' elevation={2}>
          {children}
        </Card>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should render a button if one is passed in as a prop", () => {
    const tree = renderer
      .create(
        <Card
          item={item}
          quickAddButton={button}
          orientation='column'
          elevation={2}
        >
          {children}
        </Card>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
