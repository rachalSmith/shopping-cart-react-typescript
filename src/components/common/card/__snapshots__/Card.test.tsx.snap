// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render a button if one is passed in as a prop 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1ha0qyc-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardMedia-root css-15i73xl-MuiCardMedia-root"
    role="img"
    style={
      Object {
        "backgroundImage": "url(\\"test.png\\")",
      }
    }
    title="title"
  >
    <button>
      button
    </button>
  </div>
  <div
    className="MuiCardContent-root css-8okzxw-MuiCardContent-root"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Card component should render the card image and children in a row when orientation is 'row' 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-zqrepa-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardMedia-root css-15i73xl-MuiCardMedia-root"
    role="img"
    style={
      Object {
        "backgroundImage": "url(\\"test.png\\")",
      }
    }
    title="title"
  />
  <div
    className="MuiCardContent-root css-8okzxw-MuiCardContent-root"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`Card component should render the card image and content in a column when orientation is 'column' 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1ha0qyc-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardMedia-root css-15i73xl-MuiCardMedia-root"
    role="img"
    style={
      Object {
        "backgroundImage": "url(\\"test.png\\")",
      }
    }
    title="title"
  />
  <div
    className="MuiCardContent-root css-8okzxw-MuiCardContent-root"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`should render a button if one is passed in as a prop 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1ha0qyc-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardMedia-root css-15i73xl-MuiCardMedia-root"
    role="img"
    style={
      Object {
        "backgroundImage": "url(\\"test.png\\")",
      }
    }
    title="title"
  >
    <button>
      button
    </button>
  </div>
  <div
    className="MuiCardContent-root css-8okzxw-MuiCardContent-root"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`should render the card image and children in a row when orientation is 'row' 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-zqrepa-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardMedia-root css-15i73xl-MuiCardMedia-root"
    role="img"
    style={
      Object {
        "backgroundImage": "url(\\"test.png\\")",
      }
    }
    title="title"
  />
  <div
    className="MuiCardContent-root css-8okzxw-MuiCardContent-root"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`should render the card image and content in a column when orientation is 'column' 1`] = `
<div
  className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1ha0qyc-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardMedia-root css-15i73xl-MuiCardMedia-root"
    role="img"
    style={
      Object {
        "backgroundImage": "url(\\"test.png\\")",
      }
    }
    title="title"
  />
  <div
    className="MuiCardContent-root css-8okzxw-MuiCardContent-root"
  >
    <div>
      child
    </div>
  </div>
</div>
`;
